lib/griddler/amazon_s3_ses/adapter.rb

```diff
@@ -98,10 +98,10 @@ def html_part
 
       def force_body_to_utf_8_string(message_body)
         # Fallback to an empty string literal to ensure we don't have nil.to_s which returns a frozen string.
-        # A forzen string errors out with the .force_encoding method.
+        # A frozen string errors out with the .force_encoding method.
         body_string = message_body.present? ? message_body.to_s : ""
 
-        body_string.force_encoding(Encoding::UTF_8)
+        body_string.force_encoding(Encoding::UTF_8).encode("UTF-8", invalid: :replace, replace: "")
```

Comment:
Question.  Does `.encode("UTF-8", invalid: :replace, replace: "")` makes `.force_encoding(Encoding::UTF_8)` redundant?  

I think it is okay to clean up the invalid UTF-8 character here is fine.  It makes the downstream service more stable.

Reply:
Based on this [SO post](https://stackoverflow.com/questions/21614228/what-is-the-difference-between-encode-and-force-encoding-in-ruby), it seemed like they were different and we still would want both.

Reply:
I am a bit confused by the SO post's explanation.  From the perspective of the output of the two methods (.encode vs .force_encoding), they are very similar.  

I understand this [article](https://tenderlovemaking.com/2020/01/13/guide-to-string-encoding-in-ruby.html#:~:text=The%20encode%20method%20will%20create,object%20with%20the%20new%20string.&text=Calling%20force_encoding%20mutates%20the%20original,string%20with%20a%20different%20encoding.) a little better, which explain the practical different is pretty much `.encode` creates a new copy, and `.force_encoding` mutate the variable in place.

Here is my test with `.force_encoding`, and notice the original `str_a` was mutated.
```ruby
15: [][dev] » str_a = "hello world!"
=> "hello world!"
16: [][dev] » str_a.encoding
=> #<Encoding:UTF-8>
17: [][dev] » str_b = str_a.force_encoding(Encoding::UTF_16)
=> "\x68\x65\x6C\x6C\x6F\x20\x77\x6F\x72\x6C\x64\x21"
18: [][dev] » str_a.encoding
=> #<Encoding:UTF-16 (dummy)>
19: [][dev] » str_b.encoding
=> #<Encoding:UTF-16 (dummy)>
```

Here is my test with `.encode`, and notice the original `str_a` was not changed.
```ruby
20: [][dev] » str_a = "hello world!"
=> "hello world!"
21: [][dev] » str_a.encoding
=> #<Encoding:UTF-8>
22: [][dev] » str_b = str_a.encode(Encoding::UTF_16, invalid: :replace, replace: "")
=> "\uFEFFhello world!"
23: [][dev] » str_a.encoding
=> #<Encoding:UTF-8>
24: [][dev] » str_b.encoding
=> #<Encoding:UTF-16 (dummy)>
```

I am troubled by the difference in the output of the two being different though.  

So, here is a proposed solution for you to consider too, let me know which way you prefer.

-----
Another solution as proposed by the article [here](https://tenderlovemaking.com/2020/01/13/guide-to-string-encoding-in-ruby.html#:~:text=strings%20can%20have%20the%20wrong%20encoding).

Wrong encoding from email...
```ruby
7: [][dev] » x = "Hello \x93\xfa\x96\x7b"
=> "Hello \x93\xFA\x96{"
38: [][dev] » x.force_encoding(Encoding::UTF_8)
=> "Hello \x93\xFA\x96{"
39: [][dev] » x.encoding
=> #<Encoding:UTF-8>
40: [][dev] » x.valid_encoding?
=> false
```

the .encode method:

```ruby
44: [][dev] » z = x.encode("UTF-8", invalid: :replace, replace: "")
=> "Hello {"
45: [][dev] » z.encoding
=> #<Encoding:UTF-8>
46: [][dev] » z.valid_encoding?
=> true
```

the .scrub method:

```ruby
51: [][dev] » y = x.scrub("")
=> "Hello {"
52: [][dev] » y.encoding
=> #<Encoding:UTF-8>
53: [][dev] » y.valid_encoding?
=> true
```

Reply:
I like `.scrub` better too, over calling `.encode` again. Updated in https://github.com/retailzipline/griddler-amazon_s3_ses/pull/3/commits/8ca91d577f1f591f40ef639a746c0c61ddc3ca72.